pkg/manager/member/executor_manager.go

```diff
@@ -725,7 +718,158 @@ func (m *executorMemberManager) getContainerByName(sts *appsv1.StatefulSet, cont
 	return nil
 }
 
-func (m *executorMemberManager) getVolsByName(tc *v1alpha1.TiflowCluster, sts *appsv1.StatefulSet, podName string) (map[string]*v1alpha1.StorageVolumeStatus, error) {
+func (m *executorMemberManager) syncVolsStatus(tc *v1alpha1.TiflowCluster, sts *appsv1.StatefulSet) error {
 	// todo:
-	return nil, nil
+	selector, err := metav1.LabelSelectorAsSelector(sts.Spec.Selector)
+	if err != nil {
+		klog.Errorf("converting statefulSet %s selector to metav1 selector", sts.Name)
+		return err
+	}
+
+	pvs, err := m.clientSet.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{
+		LabelSelector: selector.String(),
+	})
+
+	for _, pv := range pvs.Items {
+		tc.Status.Executor.Volumes[pv.Name] = &v1alpha1.StorageVolumeStatus{
+			Name: DefaultStorageName,
+			ObservedStorageVolumeStatus: v1alpha1.ObservedStorageVolumeStatus{
+				BoundCount:      1,
+				CurrentCount:    1,
+				ResizedCount:    1,
+				CurrentCapacity: resource.Quantity{},
+				ResizedCapacity: resource.Quantity{},
+			},
+		}
+	}
+
+	return nil
+}
+
+func (m *executorMemberManager) syncExecutorMembersStatus(tc *v1alpha1.TiflowCluster) (map[string]v1alpha1.ExecutorMember, error) {
+	if tc.Heterogeneous() && tc.Spec.Master == nil {
+		return m.syncExecutorsStatusWithHetero(tc)
+	}
+
+	return m.syncExecutorsStatusWithoutHetero(tc)
+}
+
+func (m *executorMemberManager) syncExecutorsStatusWithoutHetero(tc *v1alpha1.TiflowCluster) (map[string]v1alpha1.ExecutorMember, error) {
+	hetero := tiflowapi.Heterogeneous{}
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
+	tiflowClient := tiflowapi.GetMasterClient(m.cli, ns, tcName, "", tc.IsClusterTLSEnabled(), hetero)
+
+	// get executors info from master
+	executorsInfo, err := tiflowClient.GetExecutors()
+	if err != nil {
+		tc.Status.Executor.Synced = false
+		// get endpoints info
+		eps := &corev1.Endpoints{}
+		epErr := m.cli.Get(context.TODO(), types.NamespacedName{
+			Namespace: ns,
+			Name:      controller.TiflowMasterMemberName(tcName),
+		}, eps)
+		if epErr != nil {
+			return nil, fmt.Errorf("syncTiflowClusterStatus: failed to get endpoints %s for cluster %s/%s, err: %s, epErr %s",
+				controller.TiflowMasterMemberName(tcName), ns, tcName, err, epErr)
+		}
+		if eps != nil && len(eps.Subsets) == 0 {
+			return nil, fmt.Errorf("%s, service %s/%s has no endpoints",
+				err, ns, controller.TiflowMasterMemberName(tcName))
+		}
+		return nil, err
+	}
+
+	return syncExecutorMembers(executorsInfo)
+}
+
+func (m *executorMemberManager) syncExecutorsStatusWithHetero(tc *v1alpha1.TiflowCluster) (map[string]v1alpha1.ExecutorMember, error) {
+	ns := tc.Spec.Cluster.Namespace
+	tcName := tc.Spec.Cluster.Name
+
+	hetero := tiflowapi.Heterogeneous{}
+	if tc.Heterogeneous() && tc.Spec.Master == nil {
```

Comment:
Not elegant here will be changed later

Reply:
We are in `syncExecutorsStatusWithHetero` now, it means this if condition will always be true.

Reply:
The logic here is not always true, and does not work in a heterogeneous environment when you get the apiURL